app/controllers/session_controller.rb

```diff
@@ -424,6 +428,128 @@ def one_time_password
+        allowed_methods: challenge[:allowed_methods]
+      )
+      if user.security_keys_enabled?
+        Webauthn.stage_challenge(user, secure_session)
+        json.merge!(Webauthn.allowed_credentials(user, secure_session))
+        json[:security_keys_enabled] = true
+      else
+        json[:security_keys_enabled] = false
+      end
+    else
+      json[:error] = I18n.t(error_key)
+    end
+
+    respond_to do |format|
+      format.html do
+        store_preloaded("2fa_challenge_data", MultiJson.dump(json))
+        raise ApplicationController::RenderEmpty.new
+      end
+
+      format.json do
+        render json: json, status: status_code
+      end
+    end
+  end
+
+  def second_factor_auth_perform
+    nonce = params.require(:nonce)
+    challenge = nil
+    error_key = nil
+    status_code = 200
+    begin
+      challenge = find_second_factor_challenge(nonce)
+    rescue SecondFactor::BadChallenge => exception
+      error_key = exception.error_translation_key
+      status_code = exception.status_code
+    end
+
+    if error_key
+      json = failed_json.merge(
+        ok: false,
+        error: I18n.t(error_key),
+        reason: "challenge_not_found_or_expired"
+      )
+      render json: failed_json.merge(json), status: status_code
+      return
+    end
+
+    # no proper error messages for these cases because the only way they can
+    # happen is if someone is messing with us.
+    # the first one can only happen if someone disables a 2FA method after
+    # they're redirected to the 2fa page and then uses the same method they've
+    # disabled.
+    second_factor_method = params[:second_factor_method].to_i
+    if !current_user.valid_second_factor_method_for_user?(second_factor_method)
+      raise Discourse::InvalidAccess.new
+    end
+    # and this happens if someone tries to use a 2FA method that's not accepted
+    # for the action they're trying to perform. e.g. using backup codes to
+    # grant someone admin status.
+    if !challenge[:allowed_methods].include?(second_factor_method)
+      raise Discourse::InvalidAccess.new
+    end
+
+    if !challenge[:successful]
+      rate_limit_second_factor!(current_user)
+      second_factor_auth_result = current_user.authenticate_second_factor(params, secure_session)
+      if second_factor_auth_result.ok
+        challenge[:successful] = true
+        challenge[:generated_at] += 1.minute.to_i
+        secure_session["current_second_factor_auth_challenge"] = challenge.to_json
+      else
+        error_json = second_factor_auth_result
+          .to_h
+          .deep_symbolize_keys
+          .slice(:ok, :error, :reason)
+          .merge(failed_json)
+        render json: error_json, status: 400
+        return
+      end
+    end
+    render json: {
+      ok: true,
+      callback_method: challenge[:callback_method],
+      callback_path: challenge[:callback_path],
+      redirect_path: challenge[:redirect_path]
+    }, status: 200
+  end
+
+  if Rails.env.test?
```

Comment:
This route is meant to be an auxiliary route for the other 2 new routes in session controller, `second_factor_auth_show` and `second_factor_auth_perform`; it's not meant to replace tests for any real routes that require 2fa (if you do ctrl+f for `/session/2fa/test-action` in this PR you'll only find it used in `session_controller_spec.rb`).

For example, I need to write test case for when the `SessionController#second_factor_auth_perform` action receives a bad TOTP. In order to do this I need something that sets up a challenge in the session before I can actually send a request the `second_factor_auth_perform` action with a bad TOTP because without a challenge the `second_factor_auth_perform` would respond with 404 right away. This is where the auxiliary route comes in handy:

1. first I send a request to the auxiliary route and it takes care of setting up a challenge
2. now with the challenge in place I can a send a request to the `second_factor_auth_perform` with a bad TOTP
3. and then assert that the `second_factor_auth_perform` action responds with the right status code and error message.

See this particular flow that I've described [here](https://github.com/discourse/discourse/pull/15377/files#diff-62424ac8e3831860baf82b385e194763d10d28a036b5f0f805e947b81f4728e4R2426) and there are a few more flows that test specific edge cases in the `second_factor_auth_perform` action. Without this route, I'd need to call some other action from another controller (currently the only option would be grant admin which requires current user to b admin) in the tests of the session controller and I think that'd mix things up and cause confusion as to what the test is testing.

> we are not testing the thing that will actually happen

There is a bunch of tests in `users_controller_spec.rb` that test the whole 2fa flow for the grant admin route and the plan is to add tests like these for every route that we migrate to this new 2fa system.

Reply:
Ah thank you sorry I must have missed the ones in `users_controller_spec.rb`. Make sense now thanks for taking the time to explain :+1: